t/t2040-checkout-symlink-attr.sh

```diff
@@ -0,0 +1,54 @@
+#!/bin/sh
+
+test_description='checkout symlinks with `symlink` attribute on Windows
+
+Ensures that Git for Windows creates symlinks of the right type,
+as specified by the `symlink` attribute in `.gitattributes`.'
+
+# Tell MSYS to create native symlinks. Without this flag test-lib's
+# prerequisite detection for SYMLINKS doesn't detect the right thing.
+MSYS=winsymlinks:nativestrict && export MSYS
+
+. ./test-lib.sh
+
+if ! test_have_prereq MINGW,SYMLINKS
+then
+	skip_all='skipping $0: MinGW-only test, which requires symlink support.'
+	test_done
+fi
+
+# Adds a symlink to the index without clobbering the work tree.
+cache_symlink () {
+	sha=$(printf '%s' "$1" | git hash-object --stdin -w) &&
+	git update-index --add --cacheinfo 120000 $sha "$2"
+}
+
+# MSYS2 is very forgiving, it will resolve symlinks even if the
+# symlink type isn't correct. To make this test meaningful, try
+# them with a native, non-MSYS executable.
+cat_native () {
+	filename=$(cygpath -w "$1") &&
+	cmd.exe /c "type \"$filename\""
```

Comment:
My knee-jerk reaction was: it should be possible to verify the correct symlink type using `attrib.exe` or `fsutil.exe`, but I did not manage to find any way to do so.

The problem with the current approach is that the test case would succeed for the wrong reasons on older Windows versions (and I am fairly certain that our automated tests run on older Windows versions). But I agree that this is the best we can do for now (an alternative would be to add a helper to `t/helper/`, but that sounds like a lot of effort for little gain).

Reply:
> My knee-jerk reaction was: it should be possible to verify the correct symlink type using `attrib.exe` or `fsutil.exe`, but I did not manage to find any way to do so.

I considered that as well. It's possible to distinguish between types by running `dir` in the parent directory and checking if it gives `<SYMLINK>` or `<SYMLINKD>` as the type for the symlink.  But the parsing logic would be so complicated that I thought it would be net worse that what we have now.

It's also fairly trivial with a modern(ish) version of powershell, but I didn't want to be the one to introduce dependency of the test suite on powershell, plus I wasn't so sure that I could rely on it being being modern enough.

> The problem with the current approach is that the test case would succeed for the wrong reasons on older Windows versions (and I am fairly certain that our automated tests run on older Windows versions)

I'm not sure which by what mechanism this test would succeed on older versions of windows. The test would be skipped on versions of windows that don't support symlinks (XP and older -- on which Git doesn't work at all) or if the test suite is lacks the privilege to create symlinks (e.g. on Windows 7 without elevation or `SeCreateSymbolicLinkPrivilege` granted to standard users). But there's nothing I can do about that.

> But I agree that this is the best we can do for now (an alternative would be to add a helper to `t/helper/`, but that sounds like a lot of effort for little gain).

I'll leave it as it is, then.